### PR TITLE
Move informer event handlers to scheduler

### LINKED ISSUE #68955: [scheduler cleanup phase 1]: Move informer event handlers to pkg/scheduler
**PLEASE COORDINATE WITH SIG-SCHEDULING BEFORE WORKING ON THIS ISSUE.** We want to avoid the situation where two people accidentally start work on the same fix.
Main tracking issue: https://github.com/kubernetes/kubernetes/issues/68951

A lot of the behavior of `genericScheduler` is captured inside the informer event handlers in `pkg/scheduler/factory/factory.go`. Our code will be easier to maintain if the behavior of `genericScheduler` is not spread across different packages, so we should move all informer event handlers to be methods of `genericScheduler`.

When these functions move, we should also fix their names to be consistent. Most of them are named like `onPvAdd` and there are a few that need to be updated to match this pattern.

See also: https://github.com/kubernetes/kubernetes/issues/68954

/sig scheduling
/kind cleanup
/help

**UPDATE:**
After discussion, we decided that it is better to move the event handlers into the `k8s.io/kubernetes/pkg/scheduler` package.

### PR DESCRIPTION
**Pull request by krmayankk**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68955

```release-note
NONE
```
/kind cleanup

### CODE REVIEW

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -125,12 +172,12 @@ func (g *genericScheduler) snapshot() error {
 	// 3. snapshot ecache
 	// 4. evaludate predicates
 	// 5. stale result will be written to ecache
-	if g.equivalenceCache != nil {
-		g.equivalenceCache.Snapshot()
+	if g.equivalencePodCache != nil {
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler.go]: This seems like an old variable name. What did you base this commit on?

File: pkg/scheduler/core/generic_scheduler.go [hunk 2]
```diff
@@ -112,6 +139,26 @@ type genericScheduler struct {
 	pdbLister                algorithm.PDBLister
 	disablePreemption        bool
 	percentageOfNodesToScore int32
+	// Enable equivalence class cache
+	enableEquivalenceClassCache bool
+
+	nodeInformer                  coreinformers.NodeInformer
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler.go]: It's not clear to me that all of these informers need to be in the struct. Which ones do we need after `NewGenericScheduler()` has finished?

**Review by misterikkit — COMMENTED** [review]
Overall looks good. I haven't done a careful review of the handlers themselves.

I think the only other change I'd hope to see in this PR would be to make sure that the functions in generic_scheduler.go are organized sensibly.

**Comment by misterikkit** [comment]
This is going to need a rebase, as a small change just went in. ( #69592 )

Also, can you provide an update on the progress of this PR? It's still marked as WIP and I haven't seen responses to my previous questions. If you're stuck, reach out here or on slack.

File: pkg/scheduler/core/generic_scheduler.go (hunk 2 above)
Comment (krmayankk) [reply pkg/scheduler/core/generic_scheduler.go]: I am trying to reuse the informers passed from cmd/kube-scheduler via NewConfigFactory. I think they are not used after the NewGenericScheduler finishes. Is it ok to not use the ones passed from cmd/kube-scheduler and just a local copy to add the event handlers ?

Comment (misterikkit) [reply pkg/scheduler/core/generic_scheduler.go]: The expectation is that `NewGenericScheduler()` takes informer args for each kind that it cares about. If we add event handlers to those informers, then we might not need to keep a reference to the informer itself.

File: pkg/scheduler/factory/factory.go [hunk 3]
```diff
@@ -1222,7 +466,18 @@ func (c *configFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
 		glog.Info("Created equivalence class cache")
 	}
 
+	if c.podInformer == nil {
+		fmt.Printf("podinformer found nil\n")
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
+		// Setup volume binder
+		c.volumeBinder = volumebinder.NewVolumeBinder(c.client, c.pvcInformer, c.pvInformer, c.storageClassInformer, time.Duration(c.bindTimeoutSeconds)*time.Second)
```
Comment (krmayankk) [inline pkg/scheduler/factory/factory.go]: @misterikkit  this is the fix i added

File: pkg/scheduler/core/cache_comparer.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
+/*
```
Comment (bsalamat) [inline pkg/scheduler/core/cache_comparer.go]: Why is this added here? The comparer exists in internal/cache/comparer.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 5]
```diff
@@ -1627,6 +1627,12 @@ func (c *VolumeBindingChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMe
 		return false, nil, fmt.Errorf("node not found")
 	}
 
+	if c.binder == nil {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: are these here for debuggin?

**Review by bsalamat — COMMENTED** [review]
I am not sure about this PR, to be honest. There are parts of the scheduler logic which live outside of the generic_scheduler and manipulate the scheduler cache directly. "Assume" is one such example. This tells me that the "scheduler" should be the owner of the scheduler cache, not the "generic_scheduler". As a result the event handlers should be part of the outer "scheduler" as well.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 5 above)
Comment (krmayankk) [reply pkg/scheduler/algorithm/predicates/predicates.go]: yes debugging the test failures

**Comment by krmayankk** [comment]
@misterikkit  this is ready for review

**Review by misterikkit — CHANGES_REQUESTED** [review]
Thanks for tackling this tricky refactor! I have some comments and questions about specific changes.

File: pkg/scheduler/eventhandlers.go [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+	// NOTE: Updates must be written to scheduler cache before invalidating
+	// equivalence cache, because we could snapshot equivalence cache after the
+	// invalidation and then snapshot the cache itself. If the cache is
+	// snapshotted before updates are written, we would update equivalence
+	// cache with stale information which is based on snapshot of old cache.
+	if err := sched.config.SchedulerCache.UpdateNode(oldNode, newNode); err != nil {
+		klog.Errorf("scheduler cache UpdateNode failed: %v", err)
+	}
+
+	sched.config.SchedulingQueue.MoveAllToActiveQueue()
```
[the remark this answers is not in the document]
Comment (krmayankk) [reply pkg/scheduler/eventhandlers.go]: rebase error fixed

File: test/integration/daemonset/daemonset_test.go [hunk 7]
```diff
@@ -123,6 +123,26 @@ func setupScheduler(
...
+		informerFactory.Core().V1().PersistentVolumeClaims(),
+		informerFactory.Core().V1().ReplicationControllers(),
+		informerFactory.Apps().V1().ReplicaSets(),
+		informerFactory.Apps().V1().StatefulSets(),
+		informerFactory.Core().V1().Services(),
+		informerFactory.Policy().V1beta1().PodDisruptionBudgets(),
+		informerFactory.Storage().V1().StorageClasses(),
+	)
+
+	go func() {
```
[the remark this answers is not in the document]
Comment (krmayankk) [reply test/integration/daemonset/daemonset_test.go]: I wasnt sure about the ordering here . Do you think with event handlers moving, that wont cause any issues ?

File: pkg/scheduler/eventhandlers.go [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+	if !reflect.DeepEqual(assumedPodCopy, podCopy) {
+		return false
+	}
+	klog.V(3).Infof("Skipping pod %s/%s update", pod.Namespace, pod.Name)
+	return true
+}
+
+// AddAllEventHandlers is a helper function used in tests and in Scheduler
+// to add event handlers for various informers.
+func AddAllEventHandlers(
```
[the remark this answers is not in the document]
Comment (misterikkit) [reply pkg/scheduler/eventhandlers.go]: Sounds good - although I think @wgliang has a WIP PR to move everything to scheduler.New.

File: test/integration/daemonset/daemonset_test.go [hunk 9]
```diff
@@ -123,6 +123,26 @@ func setupScheduler(
 
 	schedulerConfig.StopEverything = stopCh
 
+	sched := scheduler.NewFromConfig(schedulerConfig)
+
+	scheduler.AddAllEventHandlers(sched,
```
[the remark this answers is not in the document]
Comment (misterikkit) [reply test/integration/daemonset/daemonset_test.go]: Sounds reasonable. Could you add a TODO in the code so we don't lose track of that?

**Review by misterikkit — CHANGES_REQUESTED** [review]
/assign @wgliang 
Can you take a look at this PR and check how it will interact with your WIP PR to move everyone to `scheduler.New`?

File: pkg/scheduler/scheduler.go [hunk 10]
```diff
@@ -55,7 +66,10 @@ const (
 // Scheduler watches for new unscheduled pods. It attempts to find
 // nodes that they fit on and writes bindings back to the api server.
 type Scheduler struct {
-	config *factory.Config
+	config         *factory.Config
+	schedulerCache schedulerinternalcache.Cache
```
[the remark this answers is not in the document]
Comment (krmayankk) [reply pkg/scheduler/scheduler.go]: removed ,they are no longer needed

File: pkg/scheduler/scheduler.go [hunk 11]
```diff
@@ -43,6 +45,15 @@ import (
...
 	"k8s.io/kubernetes/pkg/scheduler/metrics"
 	"k8s.io/kubernetes/pkg/scheduler/util"
+	"k8s.io/kubernetes/pkg/scheduler/volumebinder"
+)
+
+var (
+	serviceAffinitySet            = sets.NewString(predicates.CheckServiceAffinityPred)
+	matchInterPodAffinitySet      = sets.NewString(predicates.MatchInterPodAffinityPred)
+	generalPredicatesSets         = sets.NewString(predicates.GeneralPred)
+	noDiskConflictSet             = sets.NewString(predicates.NoDiskConflictPred)
```
[the remark this answers is not in the document]
Comment (krmayankk) [reply pkg/scheduler/scheduler.go]: actually they were no longer being used anywhere, so i removed it. They originally existed in factory.go

File: test/integration/util/util.go [hunk 12]
```diff
@@ -72,6 +72,24 @@ func StartScheduler(clientSet clientset.Interface) (factory.Configurator, Shutdo
...
+		informerFactory.Core().V1().PersistentVolumeClaims(),
+		informerFactory.Core().V1().ReplicationControllers(),
+		informerFactory.Apps().V1().ReplicaSets(),
+		informerFactory.Apps().V1().StatefulSets(),
+		informerFactory.Core().V1().Services(),
+		informerFactory.Policy().V1beta1().PodDisruptionBudgets(),
+		informerFactory.Storage().V1().StorageClasses(),
+	)
+
+	go func() {
```
Comment (misterikkit) [inline test/integration/util/util.go]: I think this goroutine is redundant with the one created in factory.NewConfigFactory

**Review by misterikkit — COMMENTED** [review]
Please address 1 new comment and squash commits, then we can get this merged. ^_^

**Review by misterikkit — APPROVED** [review]
/lgtm
:whale: :ship: 
/assign @k82cn @bsalamat 
please review for approval

**Comment by misterikkit** [comment]
@wgliang Yeah, I've tried tweaking some of the integration test code in an experimental branch, and it's a lot of work to remove `configFactory`. Seems like you need to start with the `TestContext` type.

**Review by bsalamat — COMMENTED** [review]
/approve

Thanks a lot, @krmayankk!

**Comment by krmayankk** [comment]
thanks @misterikkit @bsalamat @deads2k
